From b381628f7f5b2c31d0a33ed8908e8690faed09c7 Mon Sep 17 00:00:00 2001
From: Xianzhong <xianzhong.li@nxp.com>
Date: Fri, 14 Sep 2018 01:17:16 +0800
Subject: [PATCH 4638/5242] MGS-4200 gpu: imx: dpu-blit: fix dual prg crash
 for 8qm video

commit  bf577b94f504d10aae058f6a59b563771883dcd2 from
https://source.codeaurora.org/external/imx/linux-imx.git

prg_reg_update has kernel panic when play amphion video on 8qm:

[  161.606147] Hardware name: Freescale i.MX8QM MEK (DT)
[  161.611195] task: ffff8008f3585100 task.stack: ffff00000a738000
[  161.617112] PC is at prg_reg_update+0x20/0x28
[  161.621464] LR is at dprc_reg_update+0x38/0x58

configure and enable second prg in dprc configuration.

Signed-off-by: Xianzhong <xianzhong.li@nxp.com>
Signed-off-by: Liu Ying <victor.liu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/imx/imx8_dprc.c |   15 +++++++++++----
 1 file changed, 11 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/imx/imx8_dprc.c b/drivers/gpu/imx/imx8_dprc.c
index 0689cdd..267159c 100644
--- a/drivers/gpu/imx/imx8_dprc.c
+++ b/drivers/gpu/imx/imx8_dprc.c
@@ -402,8 +402,12 @@ void dprc_configure(struct dprc *dprc, unsigned int stream_id,
 		preq = modifier ? BYTE_64 : BYTE_1K;
 
 		dprc_write(dprc, preq, FRAME_2P_CTRL0);
-		if (dprc->sc_resource == SC_R_DC_0_BLIT1) {
-			dprc_prg_sel_configure(dprc, SC_R_DC_0_BLIT0, true);
+		if (dprc->sc_resource == SC_R_DC_0_BLIT1 ||
+		    dprc->sc_resource == SC_R_DC_1_BLIT1) {
+			dprc_prg_sel_configure(dprc,
+					dprc->sc_resource == SC_R_DC_0_BLIT1 ?
+					SC_R_DC_0_BLIT0 : SC_R_DC_1_BLIT0,
+					true);
 			prg_set_auxiliary(dprc->prgs[1]);
 			dprc->has_aux_prg = true;
 		}
@@ -411,10 +415,12 @@ void dprc_configure(struct dprc *dprc, unsigned int stream_id,
 	} else {
 		switch (dprc->sc_resource) {
 		case SC_R_DC_0_BLIT0:
-			dprc_prg_sel_configure(dprc, SC_R_DC_0_BLIT0, false);
+		case SC_R_DC_1_BLIT0:
+			dprc_prg_sel_configure(dprc, dprc->sc_resource, false);
 			prg_set_primary(dprc->prgs[0]);
 			break;
 		case SC_R_DC_0_BLIT1:
+		case SC_R_DC_1_BLIT1:
 			dprc->has_aux_prg = false;
 			break;
 		default:
@@ -694,6 +700,7 @@ bool dprc_format_supported(struct dprc *dprc, u32 format, u64 modifier)
 		case SC_R_DC_1_WARP:
 			return false;
 		case SC_R_DC_0_BLIT1:
+		case SC_R_DC_1_BLIT1:
 			return (modifier == DRM_FORMAT_MOD_NONE ||
 				modifier == DRM_FORMAT_MOD_AMPHION_TILED);
 		}
@@ -818,11 +825,11 @@ static int dprc_probe(struct platform_device *pdev)
 
 	switch (dprc->sc_resource) {
 	case SC_R_DC_0_BLIT1:
+	case SC_R_DC_1_BLIT1:
 		dprc->has_aux_prg = true;
 		/* fall-through */
 	case SC_R_DC_0_BLIT0:
 	case SC_R_DC_1_BLIT0:
-	case SC_R_DC_1_BLIT1:
 		dprc->is_blit_chan = true;
 		/* fall-through */
 	case SC_R_DC_0_FRAC0:
-- 
1.7.9.5

